gax-grpc/src/main/java/com/google/api/gax/grpc/GrpcCallContext.java

```diff
@@ -170,12 +173,28 @@ public GrpcCallContext withStreamIdleTimeout(@Nullable Duration streamIdleTimeou
     }
 
     return new GrpcCallContext(
-        channel, callOptions, streamWaitTimeout, streamIdleTimeout, channelAffinity);
+        channel, callOptions, streamWaitTimeout, streamIdleTimeout, channelAffinity, extraHeaders);
   }
 
+  @BetaApi("The surface for channel affinity is not stable yet and may change in the future.")
   public GrpcCallContext withChannelAffinity(@Nullable Integer affinity) {
     return new GrpcCallContext(
-        channel, callOptions, streamWaitTimeout, streamIdleTimeout, affinity);
+        channel, callOptions, streamWaitTimeout, streamIdleTimeout, affinity, extraHeaders);
+  }
+
+  @BetaApi("Ther surface for extra headers is not stable yet and may change in the future.")
+  public GrpcCallContext withExtraHeaders(@Nullable Metadata extraHeaders) {
```

Comment:
So header support in core gax is done in terms of `Map<String, String>` (see https://github.com/googleapis/gax-java/blob/master/gax/src/main/java/com/google/api/gax/rpc/ClientContext.java#L76 ). For any feature where the concept could be supported by http/json also, I'd like to have the surface use abstracted concepts instead of transport-specific ones (`Metadata` is grpc-specific). Accordingly, I'd like to have the ability to add headers be present in ApiCallContext. I'd like to avoid requiring external code to use `GrpcCallContext` whenever possible.

Reply:
One thing about changing to `Map<String, String>` is that `metadata` actually allows identical keys (https://grpc.io/grpc-java/javadoc/io/grpc/Metadata.html#getAll-io.grpc.Metadata.Key-), changing to use `Map<String, String>` might restrict this functionality. Are we OK with this?

Reply:
@garrettjonesgoogle On internal doc, I thought that you were in favor of accepting `Metadata`.

If you'd like a more generic version, I think we need
`withExtraHeaders(Map<String, List<String>>)`
or
`withExtraHeaders(Map<String, ? extends Collection<String>>)`
would you like that better?

@hzyi-google FYI. If we go that route, we can internally save the header with `ImmutableListMultimap`

Reply:
Haven't thought a lot into it so maybe I am wrong, but if we only want to change the surface, it's Okay to still save the header internally with metadata?

Reply:
Very true. I mostly thought that using `ImmutableListMultimap` makes it easier to guarantee immutability.

Reply:
It looks like other libraries use `Map<String, List<String>>`:

https://docs.oracle.com/javase/7/docs/api/java/net/URLConnection.html#getHeaderFields%28%29

http://square.github.io/okhttp/2.x/okhttp/com/squareup/okhttp/Headers.html#toMultimap--

I think it makes sense to transition our existing header signatures to that too (yay, they are still `@BetaApi`!)

Reply:
Regarding acceptance of `Metadata`: My prior comments on that were focused on a specific problem, and didn't take into account the larger picture of the design in GAX for handling multiple transports.

Reply:
Sounds good. I'll make the change.

Reply:
Ah, that makes sense.